talaiot/src/main/kotlin/com/cdsap/talaiot/entities/ExecutionReport.kt

```diff
@@ -0,0 +1,134 @@
+package com.cdsap.talaiot.entities
+
+data class ExecutionReport(
+    var environment: Environment = Environment(),
+    var customProperties: CustomProperties = CustomProperties(),
+    var beginMs: String? = null,
+    var endMs: String? = null,
+    var durationMs: String? = null,
+    var configurationDurationMs: String? = null,
+    var tasks: List<TaskLength>? = null,
+    var unfilteredTasks: List<TaskLength>? = null,
+    var buildId: String? = null,
+    var rootProject: String? = null,
+    var requestedTasks: String? = null,
+    var success: Boolean = false,
+    var scanLink: String? = null,
+    var buildInvocationId: String? = null
+) {
+
+    val cacheRatio: String?
+        get() = tasks?.let {
+            it.count { taskLength -> taskLength.state == TaskMessageState.FROM_CACHE } / it.size.toDouble()
+        }?.toString()
+
+    fun flattenBuildEnv(): Map<String, String> {
+        val map = mutableMapOf<String, String>()
+
+        with(environment) {
+            cacheMode?.let { map["cacheMode"] = it }
+            cachePushEnabled?.let { map["cachePushEnabled"] = it }
+            cacheUrl?.let { map["cacheUrl"] = it }
+            cacheHit?.let { map["cacheHit"] = it }
+            cacheMiss?.let { map["cacheMiss"] = it }
+            cacheStore?.let { map["cacheStore"] = it }
+
+            switches.buildCache?.let { map["switch.cache"] = it }
+            switches.buildScan?.let { map["switch.scan"] = it }
+            switches.configurationOnDemand?.let { map["switch.configurationOnDemand"] = it }
+            switches.continueOnFailure?.let { map["switch.continueOnFailure"] = it }
+            switches.daemon?.let { map["switch.daemon"] = it }
+            switches.dryRun?.let { map["switch.dryRun"] = it }
+            switches.offline?.let { map["switch.offline"] = it }
+            switches.parallel?.let { map["switch.parallel"] = it }
+            switches.refreshDependencies?.let { map["switch.refreshDependencies"] = it }
+            switches.rerunTasks?.let { map["switch.rerunTasks"] = it }
+        }
+
+        environment.osVersion?.let { map["osVersion"] = it }
+        environment.javaVmName?.let { map["javaVmName"] = it }
+        environment.cpuCount?.let { map["cpuCount"] = it }
+        environment.username?.let { map["username"] = it }
+        environment.gradleVersion?.let { map["gradleVersion"] = it }
+
+        buildId?.let { map["buildId"] = it }
+        rootProject?.let { map["rootProject"] = it }
+        requestedTasks?.let { map["requestedTasks"] = it }
+
+        //These come last to have an ability to override calculation
+        map.putAll(customProperties.properties)
+
+        return map.filter { (k, v) -> v != "undefined" }
+    }
+
+    /**
+     * This would be a lot faster if it was actually a weighted graph
```

Comment:
one question related to this comment, what should be the weight between the vertices in the graph?

Reply:
Weight would be the duration of task. So A -> B basically has the task length of A to get to the start of B.